### PR TITLE
[SPARK-36827][CORE] Improve the perf and memory usage of cleaning up stage UI data

### PR DESCRIPTION
**Pull request by gengliangwang**
### What changes were proposed in this pull request?

Improve the perf and memory usage of cleaning up stage UI data. The new code make copy of the essential fields(stage id, attempt id, completion time) to an array and determine which stage data and `RDDOperationGraphWrapper` needs to be clean based on it
### Why are the changes needed?

Fix the memory usage issue described in https://issues.apache.org/jira/browse/SPARK-36827

### Does this PR introduce _any_ user-facing change?

No

### How was this patch tested?

Add new unit test for the InMemoryStore.
Also, run a simple benchmark with 
```
    val testConf = conf.clone()
      .set(MAX_RETAINED_STAGES, 1000)

    val listener = new AppStatusListener(store, testConf, true)
    val stages = (1 to 5000).map { i =>
      val s = new StageInfo(i, 0, s"stage$i", 4, Nil, Nil, "details1",
        resourceProfileId = ResourceProfile.DEFAULT_RESOURCE_PROFILE_ID)
      s.submissionTime = Some(i.toLong)
      s
    }
    listener.onJobStart(SparkListenerJobStart(4, time, Nil, null))
    val start = System.nanoTime()
    stages.foreach { s =>
      time +=1
      s.submissionTime = Some(time)
      listener.onStageSubmitted(SparkListenerStageSubmitted(s, new Properties()))
      s.completionTime = Some(time)
      listener.onStageCompleted(SparkListenerStageCompleted(s))
    }
    println(System.nanoTime() - start)
```

Before changes:
InMemoryStore: 1.2s

After changes:
InMemoryStore: 0.23s

### CODE REVIEW

**Comment by gengliangwang** [comment]
@taroplus I was working on this yesterday. I didn't send it out because I think we can do better if we build a live priority queue in `AppStatusListener` so that Spark doesn't need to pull all the stage data out on every cleaning up.

**Comment by taroplus** [comment]
@gengliangwang sounds good, thanks !

File: core/src/main/scala/org/apache/spark/status/AppStatusListener.scala [hunk 1]
```diff
@@ -1253,44 +1254,46 @@ private[spark] class AppStatusListener(
...
+
     // As the completion time of a skipped stage is always -1, we will remove skipped stages first.
     // This is safe since the job itself contains enough information to render skipped stages in the
     // UI.
-    val view = kvstore.view(classOf[StageDataWrapper]).index("completionTime")
-    val stages = KVUtils.viewToSeq(view, countToDelete.toInt) { s =>
-      s.info.status != v1.StageStatus.ACTIVE && s.info.status != v1.StageStatus.PENDING
-    }
-
-    val stageIds = stages.map { s =>
```
Comment (gengliangwang) [inline core/src/main/scala/org/apache/spark/status/AppStatusListener.scala:1270]: I thought about keeping the original code for LevelDB here. But after investigation, I find that:
The default retained stages size is 1000, so as per
```
  private def calculateNumberToRemove(dataSize: Long, retainedSize: Long): Long = {
    if (dataSize > retainedSize) {
      math.max(retainedSize / 10L, dataSize - retainedSize)
    } else {
      0L
    }
  }
```
The `stages` here normally has a length of 100. Finding stage id inside LevelDB 100 times is not efficient, comparing to the new code. 
So I decide to make it simple and use the same code for both InMemoryStore and LevelDB.

Comment (mridulm) [reply core/src/main/scala/org/apache/spark/status/AppStatusListener.scala:1270]: I am trying to understand the last part - what is the difference w.r.t new code for finding stage id ?

Comment (mridulm) [reply core/src/main/scala/org/apache/spark/status/AppStatusListener.scala:1270]: Scratch that - did not see the attempt iteration - makes sense.

**Review by mridulm — APPROVED** [review]
Looks good to me.

**Comment by mridulm** [comment]
+CC @zhouyejoe, @thejdeep

**Comment by mridulm** [comment]
Did not notice the change was still WIP !
But the fix looks good to me - though I am not sure if it is sufficient to fix the issue

**Comment by gengliangwang** [comment]
@mridulm there is something wrong with the benchmark. I rerun the updated benchmark and find that after changes the LevelDB is slower. 
I am considering making the code changes for InMemoryStore only. So I mark it as WIP.

**Comment by gengliangwang** [comment]
@mridulm Thanks for the review. This one is ready now.
As I observe regression after applying the changes when using LevelDB, this PR is to optimize InMemoryStore only.
Now the PR is safe enough to go into Spark 3.2 as well

**Comment by mridulm** [comment]
I am guessing the minor perf difference for level db would be the sort being done after pulling all data vs pushing the sort to the store.
One option is to batch all the deletes together, instead of doing it one by one - which should make level db updates also faster ?

**Comment by gengliangwang** [comment]
> One option is to batch all the deletes together, instead of doing it one by one - which should make level db updates also faster ?

Then the same optimization can be done in the `cleanupStagesInKVStore` as well and the minor perf difference is still there. We can try that in another PR.

**Comment by mridulm** [comment]
Btw, can you share the details of the regression ?
What I had seen earlier was:
```
Before changes:
InMemoryStore: 2.8s
LevelDB: 68.9s

After changes:
InMemoryStore: 0.95s
LevelDB: 60.3s
```

That looked fine to me - did we observe a degradation in timing results after ?

**Comment by gengliangwang** [comment]
@mridulm Yes. 
At first the benchmark of job start:
```
    listener.onJobStart(SparkListenerJobStart(4, time, stages, null))
```

Later on, I change the stage info inside to `Nil`
```
    listener.onJobStart(SparkListenerJobStart(4, time, Nil, null))
```
and the difference for LevelDB is
before changes: 2.5s
after changes: 3s

File: core/src/main/scala/org/apache/spark/status/AppStatusListener.scala [hunk 2]
```diff
@@ -1253,12 +1254,47 @@ private[spark] class AppStatusListener(
...
-      return
+  private case class StageCompletionTime(
+      stageId: Int,
+      attemptId: Int,
+      completionTime: Long)
+
+  private def cleanupStagesWithInMemoryStore(countToDelete: Long): Seq[Array[Int]] = {
+    val stageArray = new ArrayBuffer[StageCompletionTime]()
+    val stageDataCount = new mutable.HashMap[Int, Int]()
+    kvstore.view(classOf[StageDataWrapper]).forEach { s =>
```
Comment (cloud-fan) [inline core/src/main/scala/org/apache/spark/status/AppStatusListener.scala:1265]: I think `KVUtils.viewToSeq(view, countToDelete.toInt)` is better and cleaner if the in-memory store implements it with top-n algorithm. We can do it in followups.

**Review by cloud-fan — APPROVED** [review]

**Review by cloud-fan — APPROVED** [review]

File: core/src/main/scala/org/apache/spark/status/AppStatusListener.scala [hunk 3]
```diff
@@ -1253,12 +1254,47 @@ private[spark] class AppStatusListener(
...
+      // Here we keep track of the total number of StageDataWrapper entries for each stage id.
+      // This will be used in cleaning up the RDDOperationGraphWrapper data.
+      if (stageDataCount.contains(s.info.stageId)) {
+        stageDataCount(s.info.stageId) += 1
+      } else {
+        stageDataCount(s.info.stageId) = 1
+      }
```
Comment (Ngone51) [inline core/src/main/scala/org/apache/spark/status/AppStatusListener.scala:1272]: this's only needed for completed stages?

**Review by Ngone51 — APPROVED** [review]
LGTM

**Comment by gengliangwang** [comment]
@taroplus @mridulm @Ngone51 @cloud-fan Thanks for the review.
Merging to master/3.2

File: core/src/main/scala/org/apache/spark/status/AppStatusListener.scala [hunk 4]
```diff
@@ -1294,6 +1330,22 @@ private[spark] class AppStatusListener(
...
+    val countToDelete = calculateNumberToRemove(count, conf.get(MAX_RETAINED_STAGES))
+    if (countToDelete <= 0L) {
+      return
+    }
+
+    // SPARK-36827: For better performance and avoiding OOM, here we use a optimized method for
+    //              cleaning the StageDataWrapper and RDDOperationGraphWrapper data if Spark is
+    //              using InMemoryStore.
+    val stageIds = if (kvstore.usingInMemoryStore) {
+      cleanupStagesWithInMemoryStore(countToDelete)
```
Comment (srowen) [inline core/src/main/scala/org/apache/spark/status/AppStatusListener.scala:1345]: Could this be pushed down to the in memory store? vs special casing it here. This doesn't seem very object-oriented

Comment (dongjoon-hyun) [reply core/src/main/scala/org/apache/spark/status/AppStatusListener.scala:1345]: +1 for @srowen 's comment.

File: core/src/main/scala/org/apache/spark/status/AppStatusListener.scala [hunk 5]
```diff
@@ -1294,6 +1330,22 @@ private[spark] class AppStatusListener(
...
     }
+  }
+
+  private def cleanupStages(count: Long): Unit = {
+    val countToDelete = calculateNumberToRemove(count, conf.get(MAX_RETAINED_STAGES))
+    if (countToDelete <= 0L) {
+      return
+    }
+
+    // SPARK-36827: For better performance and avoiding OOM, here we use a optimized method for
```
Comment (dongjoon-hyun) [inline core/src/main/scala/org/apache/spark/status/AppStatusListener.scala:1341]: I believe we can remove `SPARK-36827: ` and the below indentations inside this comment.

**Comment by dongjoon-hyun** [comment]
+1, LGTM, too. Thank you, @gengliangwang and all.

File: core/src/main/scala/org/apache/spark/status/AppStatusListener.scala (hunk 4 above)
Comment (cloud-fan) [reply core/src/main/scala/org/apache/spark/status/AppStatusListener.scala:1345]: maybe we need a more high-level API in the kv store to clean up stages?

**Comment by zhouyejoe** [comment]
@gengliangwang 
QQ regarding the regression test:

> the difference for LevelDB is
> before changes: 2.5s
> after changes: 3s

How many stages within your test log file for SHS? Or will this performance regression get amplified if there are more stages within the application?

File: core/src/main/scala/org/apache/spark/status/AppStatusListener.scala (hunk 4 above)
Comment (gengliangwang) [reply core/src/main/scala/org/apache/spark/status/AppStatusListener.scala:1345]: @srowen +1. I think we can have an enhanced version of in-memory `ElementTrackingStore`.
I was trying to merge this one in 3.2.0 so this is a temporary solution.

**Comment by gengliangwang** [comment]
@zhouyejoe I use the benchmark code in the PR description, which is runnable under `AppStatusListenerSuite`.
I tried 3000 and 5000 stages the result "after changes" is always slightly slow. 
When the number of stages reaches 1001, Spark will delete the oldest 100 stages UI data. I guess after changes reading 1001 stages from LevelDB is slow. Before changes Spark only reads 100 stages from LevelDB.

**Pull request closed without merging**